### PR TITLE
Add operator-ci to execute integration tests for PRs, nightly, and releases

### PR DESCRIPTION
**Pull request by arilivigni**
- based on devconsole-operator

### CODE REVIEW

**Comment by arilivigni** [comment]
@nikhil-thomas this needs to be merged becausehttps://github.com/openshift/release/pull/3964  is dependent on it

File: make/dev.mk [hunk 1]
```diff
@@ -0,0 +1,106 @@
+ifndef DEV_MK
+DEV_MK:=# Prevent repeated "-include".
+
+include ./make/verbose.mk
+include ./make/git.mk
+
+DOCKER_REPO?=quay.io/openshiftio
```
Comment (sthaha) [inline make/dev.mk]: the docker repo needs to change

File: make/dev.mk [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+IMAGE_NAME?=tektoncd-pipeline-operator
+REGISTRY_URI=quay.io
+
+DEVCONSOLE_OPERATOR_IMAGE?=quay.io/redhat-developers/tektoncd-pipeline-operator
```
Comment (sthaha) [inline make/dev.mk]: do we need a DEVCONSOLE here?

File: make/dev.mk [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+## Run Operator locally
+local: deploy-rbac build deploy-crd
+	$(Q)-oc new-project $(LOCAL_TEST_NAMESPACE)
+	$(Q)operator-sdk up local --namespace=$(APP_NAMESPACE)
+
+.PHONY: deploy-rbac
+## Setup service account and deploy RBAC
+deploy-rbac:
+	$(Q)-oc login -u system:admin
+	$(Q)-oc create -f deploy/service_account.yaml
```
Comment (sthaha) [inline make/dev.mk]: how about we just `oc create -f deploy/` instead?

File: make/dev.mk [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	$(Q)-oc create -f deploy/role.yaml
+	$(Q)-oc create -f deploy/role_binding.yaml
+
+.PHONY: deploy-crd
+## Deploy CRD
+deploy-crd:
+	$(Q)-oc apply -f deploy/crds/devconsole_v1alpha1_component_crd.yaml
```
Comment (sthaha) [inline make/dev.mk]: those are not the crds that operator owns

File: make/dev.mk [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	@oc delete -f ./deploy/service_account.yaml || true
+	@echo "Deleting Custom Resource Definitions..."
+	@oc delete -f ./deploy/crds/devconsole_v1alpha1_gitsource_crd.yaml || true
+
+.PHONY: deploy-operator
+deploy-operator: build build-operator-image deploy-operator-only
+
+.PHONY: minishift-start
+minishift-start:
+	minishift start --cpus 4 --memory 8GB
```
Comment (sthaha) [inline make/dev.mk]:
```
minikube start --cpus 4 --memory 8GB \ 
  --extra-config=apiserver.enable-admission-plugins="LimitRanger,NamespaceExists,NamespaceLifecycle,ResourceQuota,ServiceAccount,DefaultStorageClass,MutatingAdmissionWebhook" \
 --extra-config=apiserver.service-node-port-range=80-32767
```

File: make/docker.mk [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+# If running in Jenkins we don't allow for interactively running the container
+DOCKER_RUN_INTERACTIVE_SWITCH = -i
+ifneq ($(BUILD_TAG),)
+	DOCKER_RUN_INTERACTIVE_SWITCH =
+endif
+
+.PHONY: docker-image-deploy
+## Build the docker image that can be deployed (only contains bare operator)
+docker-image-deploy: Dockerfile
```
Comment (sthaha) [inline make/docker.mk]: IMHO we should use ` operator-sdk build ` to build the image

File: make/dev.mk (hunk 2 above)
Comment (arilivigni) [reply make/dev.mk]: Sorry I will update that to reflect our operator

File: make/docker.mk (hunk 6 above)
Comment (arilivigni) [reply make/docker.mk]: I was told to model the devconsole-operator by @nikhil-thomas do we want to change this?

File: make/dev.mk (hunk 4 above)
Comment (arilivigni) [reply make/dev.mk]: Sorry I will swamp that out for the correct ones for tekton

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# It's necessary to set this because some environments don't link sh -> bash.
+SHELL := /bin/bash
```
Comment (vdemeester) [inline Makefile]: :sob: this will fail on some system (Nixos, OSX, …)

File: make/dev.mk [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	$(Q)-oc apply -f examples/devconsole_v1alpha1_gitsourceanalysis_cr.yaml
+
+.PHONY: build-operator-image
+## Build and create the operator container image
+build-operator-image: ./vendor
+	operator-sdk build $(DEVCONSOLE_OPERATOR_IMAGE):$(TAG)
+
+.PHONY: push-operator-image
+## Push the operator container image to a container registry
+push-operator-image: build-operator-image
```
Comment (vdemeester) [inline make/dev.mk]: Do we need those here for now ? (aka we will/should use openshift/release CI capabilities here, promoting and mirroring image from the ci registry to quay)

File: make/dev.mk [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	@echo "Deleting Custom Resource Definitions..."
+	@oc delete -f ./deploy/crds/devconsole_v1alpha1_gitsource_crd.yaml || true
+
+.PHONY: deploy-operator
+deploy-operator: build build-operator-image deploy-operator-only
+
+.PHONY: minishift-start
+minishift-start:
+	minishift start --cpus 4 --memory 8GB
+	-eval `minishift docker-env` && oc login -u system:admin
```
Comment (vdemeester) [inline make/dev.mk]: `s/minishift/minikube` and remove `oc login`

File: make/docker.mk (hunk 6 above)
Comment (vdemeester) [reply make/docker.mk]: I am with @sthaha (especially as, some of us don't even have `docker` installed :sweat_smile: )

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	$(Q)go clean ${X_FLAG} ./...
+
+./vendor: Gopkg.toml Gopkg.lock
+	$(Q)dep ensure ${V_FLAG} -vendor-only
+
+./out/operator: ./vendor $(shell find . -path ./vendor -prune -o -name '*.go' -print)
+	#$(Q)operator-sdk generate k8s
+	$(Q)CGO_ENABLED=0 GOARCH=amd64 GOOS=linux \
+		go build ${V_FLAG} \
+		-ldflags "-X ${GO_PACKAGE_PATH}/cmd/manager.Commit=${GIT_COMMIT_ID} -X ${GO_PACKAGE_PATH}/cmd/manager.BuildTime=${BUILD_TIME}" \
```
Comment (vdemeester) [inline Makefile]: Do we have those variable in `cmd/manager` ? (@nikhil-thomas @sthaha )

File: openshift-ci/Dockerfile.tools [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    make \
+    procps-ng \
+    tar \
+    wget \
+    which \
+    bc \
+    kubectl \
+    yamllint \
+    python36-virtualenv \
+    chromium \
```
Comment (vdemeester) [inline openshift-ci/Dockerfile.tools]: We don't need those (no UI for this operator)…

File: openshift-ci/README.md [hunk 12]
```diff
@@ -0,0 +1,214 @@
+# TektonCD Pipeline Operator Continuous Integration
+
+TektonCD Pipeline operator uses [OpenShift CI][openshift-ci] for continuous
+integration.  The Openshift CI is built using
+[CI-Operator][ci-operator].  The TektonCD Pipeline Operator specific
+configuration is located here: http://bit.ly/304kpIo
```
Comment (vdemeester) [inline openshift-ci/README.md]: The link will need to be updated

File: openshift-ci/README.md [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+### tektoncd-pipeline-operator
+
+Thee CentOS 7 image (`os`) is going to use as the base image for
+creating `tektoncd-pipeline-operator` image.  The operator binary available
+inside `bin` container image is copied over here.  The Dockerfile used
+is available here: `openshift-ci/Dockerfile.deploy`
+
+The image produced can be pulled from here:
+`registry.svc.ci.openshift.org/${OPENSHIFT_BUILD_NAMESPACE}/stable:tektoncd-pipeline-operator`
```
Comment (vdemeester) [inline openshift-ci/README.md]: This depends on the `openshift/release` configuration. Is it gonna be `stable` for the name ?

File: make/docker.mk (hunk 6 above)
Comment (arilivigni) [reply make/docker.mk]: Ok I will use operator-sdk as much as possible

File: openshift-ci/README.md (hunk 13 above)
Comment (arilivigni) [reply openshift-ci/README.md]: I will double check

File: Makefile (hunk 10 above)
Comment (vdemeester) [reply Makefile]: answering my own question : https://github.com/openshift/tektoncd-pipeline-operator/blob/master/cmd/manager/main.go doesn't have `Commit` or `BuildTime` variable, so we can remove those `ldflags`

File: make/docker.mk (hunk 6 above)
Comment (nikhil-thomas) [reply make/docker.mk]: the only reason not to use `operator-sdk build` is a need for supplying go build flags.
it is a known issue https://github.com/operator-framework/operator-sdk/issues/1435

in devconsole operator they want to tag their binary with commit-hash build time
here https://github.com/operator-framework/operator-sdk/issues/1435

as we are not doing it right now, we can use `operator-sdk build`

File: make/test.mk [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+	$(Q)sed -e "s,REPLACE_IMAGE,registry.svc.ci.openshift.org/${OPENSHIFT_BUILD_NAMESPACE}/stable:tektoncd-pipeline-operator-registry," ./test/e2e/catalog_source_OS4.yaml | oc apply -f -
+	$(Q)oc apply -f ./test/e2e/subscription_OS4.yaml
+	$(eval DEPLOYED_NAMESPACE := openshift-operators)
+	$(Q)./hack/check-crds.sh
+	$(Q)operator-sdk test local ./test/e2e --no-setup --go-test-flags "-v -timeout=15m"
+
+.PHONY: test-e2e-ci
+test-e2e-ci: get-test-namespace ./vendor
+	$(Q)oc new-project $(TEST_NAMESPACE)
+	$(Q)-oc apply -f ./deploy/crds/devconsole_v1alpha1_component_crd.yaml
```
Comment (nikhil-thomas) [inline make/test.mk]: we don't these crds.
we can either change it to our CRD or remove the targets which are not being tested at present

File: make/test.mk [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+build-image-local: e2e-setup-local
+	eval $$(minishift docker-env) && operator-sdk build $(shell minishift openshift registry)/$(TEST_NAMESPACE)/tektoncd-pipeline-operator
+
+.PHONY: test-e2e-local
+test-e2e-local: build-image-local
+	$(eval DEPLOYED_NAMESPACE := $(TEST_NAMESPACE))
+	$(Q)-oc login -u system:admin
+	$(Q)-oc project $(TEST_NAMESPACE)
+	$(Q)-oc create -f ./deploy/crds/devconsole_v1alpha1_component_crd.yaml
+	$(Q)-oc create -f ./deploy/crds/devconsole_v1alpha1_gitsource_crd.yaml
```
Comment (nikhil-thomas) [inline make/test.mk]: remove devconsole CRDs

**Pull request merged**